### ISSUE TITLE
linear-gradient() / rgb  is removed from used CSS in some cases

### ISSUE THREAD

**Issue by Mai-Saad** [wp-media/wp-rocket#4493]
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version => yes
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
linear-gradient is removed in the following cases:

`background-color: linear-gradient(to bottom, rgb(100%, 0%, 0%), rgb(110%, 0%, 0%), rgb(255, 0, 153.0), rgb(255, 0, 153.6, 1));`

`background-image: linear-gradient(to bottom, rgb(2.55e2, 0e0, 1.53e2, 1e2), rgba(51, 170, 51, .7) , rgb(255, 0, 153.0), rgb(255, 0, 153.6, 1), rgb(146.064 107.457 131.223));`

`background-image: linear-gradient(to bottom, hsl(270,60%,70%), hsl(270 60% 70%) , hsl(270deg, 60%, 70%), hsl(4.71239rad, 60%, 70%), hsl(.75turn, 60%, 70%));`

`background-image: linear-gradient(to bottom, hsl(270, 60%, 50%, .15), hsl(270, 60%, 50%, 15%) , hsl(270 60% 50% / .15), hsl(270 60% 50% / 15%));`

`background-image: linear-gradient(to bottom, hsla(240, 100%, 50%, .05), hsla(240 100% 50% / .05) , hsla(240 100% 50% / 5%), rgb(255,-10,0), rgba(100%,0%,0%,1));`

**Expected behavior**
linear-gradient not removed

**Additional context**
- This is removed from used CSS too `background-color: rgb(0 0 0 / 10.5%);` => this is impacting as well that we can't safelist this selector `.selector-a3` (I think the problem here is mainly decimal in % as integer % working fine)

```
.selector-a3{
	-webkit-box-shadow: 0 1px 6px 0 rgb(0 0 0 / 10.5%);
	box-shadow: 0 1px 6px 0 rgb(0 0 0 / 10.5%);
}

```

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

**Comment by vmanthos** [comment]
Related ticket: https://secure.helpscout.net/conversation/1690714535/306573?folderId=2135277

The following rule is removed and the background image isn't displayed:
```CSS
div.et_pb_section.et_pb_section_2 {
    background-image: linear-gradient(
180deg,#ffffff 20%,rgba(255,255,255,0) 80%),url(https://example.com/wp-content/uploads/2019/08/welkom-ophangen-1.jpg)!important;
}
```

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1684637888/305409/

```CSS
.ultp-block-f27b18 .ultp-block-content .ultp-block-title a {
                color: rgba(0,0,0,1);
            }
```

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1683501626/305165?folderId=2135277

The `background-color` was removed from the following rule:
```CSS
i.fb-icon-element.fontawesome-icon.fb-icon-element-1 {
    color: #000000;
    background-color: rgba(244,153,34,0);
    border-color: #000000;
 }
```

**Comment by vmanthos** [comment]
Related ticket: https://secure.helpscout.net/conversation/1705216128/309548?folderId=2135277

The `background-color` was removed from the following:
```CSS
body.kc-css-system .kc-css-981754 a {
                background-color: rgba(255, 255, 255, 0);
                text-align: left;
                padding: 3px 3px 3px 3px;
                margin-right: 5px;
            }
```

**Comment by iCaspar** [comment]
Several items in the OP, including (at least) the following:
- Parser library seems not able to accept decimal values in colors.
- Parser library seems not able to accept scientific notation values in colors
- For hsl(a), parser library cannot accept level 4 color spec allowing spaces w/o comma separators
- Parser library has [a bug](https://github.com/sabberworm/PHP-CSS-Parser/blob/3537a7096e8dd5750c032e7429042a4d3d29aabe/src/Value/Size.php#L27) where it recognizes `turn` unit as `turns` instead

Additional comments from subsequent tickets imply that the use of 0 or 1 as the alpha value may be problematic.

In the case of the hsl(a), we can create a preg_replace, similar to the current one handling rgba with spacing. [S]

In the case of `turns` we could string replace any turn unit to use `turns` instead (but this still may be problematic in the generated usedCSS and need to be changed back to `turn` on the other end of the parsing). [S-M] (will need to try first string replacement, then see if we need the reverse after conversion).

For scientific notation, we will need some kind of preg_match to identify, then send to a converter, then truncate/round the fractional part of the result (see next). [M] Need to find (1) regex, and (2) converter.

For decimals, the [level 3 color specification](https://drafts.csswg.org/css-color-3/#rgb-color) implies that fractional values are not really valid, as everything must be convertible to an integer 0-255 corresponding to the bit-wise representation in the #rrggbb space. The alpha component is different, as it is by definition a fractional value. Truncating or rounding decimals is probably the best solution. [S-M] Identify the decimals in all the places via regex. Also, the decimal value in an alpha expressed as % should work, but will need to be converted from % to decimal. 🤯 

For alphas of 0/1 -- we need some further investigation at this point, as this should be recognized.

**Comment by piotrbak** [comment]
Since this is going to to be [S] + [M] + 2* [S-M], we're going to block this and check if the problem is fixed within the new RUCSS Approach.

**Comment by piotrbak** [comment]
Fixed with 3.11
